lib/internal/modules/esm/loader.js

```diff
@@ -408,41 +452,97 @@ class ESMLoader {
 
     const namespaces = await PromiseAll(new SafeArrayIterator(jobs));
 
-    return wasArr ?
-      namespaces :
-      namespaces[0];
+    if (!wasArr) return namespaces[0];
+
+    const namespaceMap = new SafeMap();
+
+    for (let i = 0; i < count; i++) namespaceMap.set(specifiers[i], namespaces[i]);
+
+    return namespaceMap;
   }
 
   /**
    * Provide source that is understood by one of Node's translators.
    *
-   * The internals of this WILL change when chaining is implemented,
-   * depending on the resolution/consensus from #36954
-   * @param {string} url The URL/path of the module to be loaded
+   * @param {URL['href']} url The URL/path of the module to be loaded
    * @param {object} context Metadata about the module
-   * @returns {object}
+   * @returns {{ format: ModuleFormat, source: ModuleSource }}
    */
   async load(url, context = {}) {
-    const defaultLoader = this.#loaders[0];
+    const loaders = this.#loaders.entries();
+    let {
+      0: loaderFilePath,
+      1: loader,
+    } = loaders.next().value;
+    let chainFinished = this.#loaders.size === 1;
+
+    function next(nextUrl, ctx = context) {
+      const {
+        done,
+        value,
+      } = loaders.next();
+      ({
+        0: loaderFilePath,
+        1: loader,
+      } = value);
+
+      if (done || loader === defaultLoad) chainFinished = true;
+
+      const hookErrIdentifier = `${loaderFilePath} "load"`;
+
+      if (typeof nextUrl !== 'string') { // non-strings can be coerced to a url string
+        throw new ERR_INVALID_ARG_TYPE(
+          `${hookErrIdentifier} next(url)`,
+          'a url string',
+          nextUrl,
+        );
+      }
+
+      try {
+        new URL(nextUrl);
+      } catch (err) { // eslint-disable-line no-unused-vars
+        throw new ERR_INVALID_ARG_VALUE(
+          `${hookErrIdentifier} next(url)`,
+          nextUrl,
+          'should be a url string',
+        );
+      }
```

Comment:
@guybedford you had previously raised a perf concern when I added a similar check in #41446. This new addition has much more impact as it is run for every hook in the load chain (whereas the 41446 check was only once per import).

I can't think of a better way to validate the url (`nextUrl`) being passed from one `load` hook to the next, but I do think we need to ensure it is validated.

I'd be grateful for any suggestions 🙂

Reply:
ping @guybedford